File: Mage/Map/Cache/XYZDirectoryCacheOverlay.m
//
//  XYZDirectoryCacheOverlay.m
//  MAGE
//
//  Created by Brian Osborn on 12/18/15.
//  Copyright © 2015 National Geospatial Intelligence Agency. All rights reserved.
//

#import "XYZDirectoryCacheOverlay.h"

@interface XYZDirectoryCacheOverlay ()

@property (strong, nonatomic) NSString * directory;

@end

@implementation XYZDirectoryCacheOverlay

-(instancetype) initWithName: (NSString *) name andDirectory: (NSString *) directory{
    self = [super initWithName:name andType:XYZ_DIRECTORY andSupportsChildren:false];
    if(self){
        self.directory = directory;
        self.tileCount = 0;
        self.minZoom = 100;
        self.maxZoom = -1;
        
        NSArray<NSString *> *zooms = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:directory error:nil];
        for (NSString *zoom in zooms) {
            self.minZoom = MIN(self.minZoom, zoom.intValue);
            self.maxZoom = MAX(self.maxZoom, zoom.intValue);
            NSString *zoomPath = [directory stringByAppendingPathComponent:zoom];
            for (NSString *x in [[NSFileManager defaultManager] contentsOfDirectoryAtPath:zoomPath error:nil]) {
                NSString *xPath = [zoomPath stringByAppendingPathComponent:x];
                self.tileCount += (int)[[NSFileManager defaultManager] contentsOfDirectoryAtPath:xPath error:nil].count;
            }
        }
    }
    return self;
}

-(void) removeFromMap: (MKMapView *) mapView{
    if(self.tileOverlay != nil){
        [mapView removeOverlay:self.tileOverlay];
        self.tileOverlay = nil;
    }
}

-(NSString *) getIconImageName{
    return @"layers";
}

-(NSString *) getDirectory{
    return self.directory;
}

- (NSString *) getInfo {
    return [NSString stringWithFormat:@"%d tiles, zoom: %d - %d", self.tileCount, self.minZoom, self.maxZoom];
}

@end

File: Mage/Map/Cache/XYZDirectoryCacheOverlay.swift
//
//  XYZDirectoryCacheOverlay.m
//  MAGE
//
//  Created by Brian Osborn on 12/18/15.
//  Copyright © 2015 National Geospatial Intelligence Agency. All rights reserved.
//

class XYZDirectoryCacheOverlay: CacheOverlay {
    var tileOverlay: MKTileOverlay?
    var minZoom: Int
    var maxZoom: Int
    var tileCount: Int
    var directory: String?
    
    init(name: String, directory: String) {
        self.directory = directory
        tileCount = 0
        minZoom = 100
        maxZoom = -1
        super.init(name: name, type: CacheOverlayType.XYZ_DIRECTORY, supportsChildren: false)
        self.iconImageName = "layers"
        
        let zooms = try? FileManager.default.contentsOfDirectory(atPath: directory)
        for zoom in zooms ?? [] {
            minZoom = min(minZoom, Int(zoom) ?? 100)
            maxZoom = max(maxZoom, Int(zoom) ?? -1)
            let zoomPath = (directory as NSString).appendingPathComponent(zoom)
            let xDirectories = try? FileManager.default.contentsOfDirectory(atPath: zoomPath)
            for xDirectory in xDirectories ?? [] {
                let xPath = (zoomPath as NSString).appendingPathComponent(xDirectory)
                let yDirectories = try? FileManager.default.contentsOfDirectory(atPath: xPath)
                for yDirectory in yDirectories ?? [] {
                    let yPath = (xPath as NSString).appendingPathComponent(yDirectory)
                    tileCount += (try? FileManager.default.contentsOfDirectory(atPath: yPath).count) ?? 0
                }
            }
        }
    }
    
    override func removeFromMap(mapView: MKMapView) {
        if let tileOverlay = tileOverlay {
            mapView.removeOverlay(tileOverlay)
            self.tileOverlay = nil
        }
    }
    
    override func getInfo() -> String? {
        "\(self.tileCount) tiles, zoom: \(self.minZoom) - \(self.maxZoom)"
    }
}
